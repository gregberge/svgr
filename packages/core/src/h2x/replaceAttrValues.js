const replaceAttrValues = (attributes = {}) => () => ({
  visitor: {
    JSXAttribute: {
      enter(path) {
        Object.keys(attributes).forEach(key => {
          if (path.node.value === key) {
            path.node.value = attributes[key]
          }
        })
      },
    },
  },
})

export default replaceAttrValues
