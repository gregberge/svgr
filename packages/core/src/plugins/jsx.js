import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
