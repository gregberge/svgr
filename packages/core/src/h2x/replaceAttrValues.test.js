import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import replaceAttrValues from './replaceAttrValues'

describe('replaceAttrValues', () => {
  it('should replace an attribute value by another', () => {
    const result = transform(
      `
      <svg>
        <g stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape"></path>
        </g>
      </svg>
    `,
      { plugins: [jsx, replaceAttrValues({ '#063855': 'currentColor' })] },
    )

    expect(result).toMatchSnapshot()
  })
})
