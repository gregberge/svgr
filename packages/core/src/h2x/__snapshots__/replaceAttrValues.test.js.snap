// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceAttrValues should replace an attribute value by another 1`] = `
"<svg>
  <g stroke=\\"currentColor\\" strokeWidth={2}>
    <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
  </g>
</svg>
"
`;
